libraries/SoftwareSerial/SoftwareSerial.h

```diff
@@ -0,0 +1,471 @@
+   // Precalculate the various delays
+   // Calculate the distance between bit in micro seconds
+   uint32_t bit_delay = (float(1) / speed) * 1000000;
+
+   _tx_delay_us = bit_delay;
+#if (_SS_TX_ONLY == 0)
+   // Wait 1/2 bit - 2 micro seconds (time for interrupt to be served)
+   _rx_delay_centering_us = (bit_delay / 2) - 2;
+   // Wait 1 bit - 2 micro seconds (time in each loop iteration)
+   _rx_delay_intrabit_us = bit_delay - 1;  // 2
+   // Wait 1 bit (the stop one)
+   _rx_delay_stopbit_us = bit_delay;
+#endif
+   delayMicroseconds(_tx_delay_us);
+
+   listen();
+}
+
+bool SoftwareSerial::listen()
+{
+#if (_SS_TX_ONLY == 0)
+   if (!_rx_delay_stopbit_us) return false;
+
+   if (active_object != this)
+   {
+      if (active_object) active_object->stopListening();
+
+      _buffer_overflow     = false;
+      _receive_buffer_head = _receive_buffer_tail = 0;
+      active_object                               = this;
+
+      if (_inverse_logic)
+      {
+         // Start bit high
+         _intMask = attachInterrupt(_receivePin, handle_interrupt, RISING);
+      }
+      else
+      {
+         // Start bit low
+         _intMask = attachInterrupt(_receivePin, handle_interrupt, FALLING);
+      }
+      return true;
+   }
+#endif
+   return false;
+}
+
+bool SoftwareSerial::stopListening()
+{
+#if (_SS_TX_ONLY == 0)
+   if (active_object == this)
+   {
+      detachInterrupt(_receivePin);
+      active_object = NULL;
+      return true;
+   }
+#endif
+   return false;
+}
+
+void SoftwareSerial::end() { stopListening(); }
+
+int SoftwareSerial::read()
+{
+#if (_SS_TX_ONLY == 1)
+   this->println(F("read() is invalid - _SS_TX_ONLY is defined."));
+#else
+   if (isListening() == 0)
+   {
+      return -1;
+   }
+
+   // Empty buffer?
+   if (_receive_buffer_head == _receive_buffer_tail)
+   {
+      return -1;
+   }
+
+   // Read from "head"
+   uint8_t d            = _receive_buffer[_receive_buffer_head];  // grab next byte
+   _receive_buffer_head = (_receive_buffer_head + 1) % _SS_MAX_RX_BUFF;
+   return d;
+#endif
+}
+
+int SoftwareSerial::available()
+{
+#if (_SS_TX_ONLY == 0)
+   if (isListening() == 0)
+      return 0;
+   else
+      return (_receive_buffer_tail + _SS_MAX_RX_BUFF - _receive_buffer_head) % _SS_MAX_RX_BUFF;
+#else
+   return 0;
+#endif
+}
+
+size_t SoftwareSerial::write(uint8_t b)
+{
+   if (_tx_delay_us == 0)
+   {
+      setWriteError();
+      return 0;
+   }
+
+   // By declaring these as local variables, the compiler will put them
+   // in registers _before_ disabling interrupts and entering the
+   // critical timing sections below, which makes it a lot easier to
+   // verify the cycle timings
+   volatile uint32_t* reg = _transmitPortRegister;
+   uint32_t reg_mask      = _transmitBitMask;
+   uint32_t inv_mask      = ~_transmitBitMask;
+   bool inv               = _inverse_logic;
+   uint16_t delay         = _tx_delay_us;
+
+   if (inv) b = ~b;
+
+// turn off interrupts for a clean txmit
+#ifndef _SS_TX_ONLY
```

Comment:
`_SS_TX_ONLY` is always defined, either by the user or as `0` at the top of the file.
There is a few of these around, search for `ifndef` or `ifdef`.